sw/airborne/modules/relativelocalizationfilter/relativelocalizationfilter.c

```diff
@@ -0,0 +1,203 @@
+ *
+ * This file is part of paparazzi
+ *
+ * paparazzi is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file "modules/relativelocalizationfilter/relativelocalizationfilter.c"
+ * @author Mario Coppola
+ * Relative Localization Filter for collision avoidance between drones
+ */
+
+#include "relativelocalizationfilter.h"
+#include "subsystems/datalink/telemetry.h"
+#include "modules/multi/traffic_info.h"
+#include "modules/stdma/stdma.h"
+#include "subsystems/datalink/bluegiga.h"
+
+#include "pprzlink/pprz_transport.h"
+#include "subsystems/datalink/datalink.h"
+#include "subsystems/datalink/downlink.h"
+
+#ifndef NUAVS
+#define NUAVS 5				// Maximum expected number of drones
+#endif
+
+#ifndef INS_INT_VEL_ID
+#define INS_INT_VEL_ID ABI_BROADCAST
+#endif
+
+ekf_filter ekf[NUAVS-1]; 	// EKF structure
+btmodel model[NUAVS-1];  	// Bluetooth model structure 
+int IDarray[NUAVS-1]; 		// Array of IDs of other MAVs
+int8_t srcstrength[NUAVS-1];// Source strength
+uint32_t now_ts[NUAVS-1]; 	// Time of last received message from each MAV
+int nf;						// Number of filters registered
+float RSSIarray[NUAVS-1];	// Recorded RSSI values (so they can all be sent)
+
+static abi_event rssi_ev;
+static void bluetoothmsg_cb(uint8_t sender_id __attribute__((unused)), 
+	uint8_t ac_id, int8_t source_strength, int8_t rssi);
+
+static void bluetoothmsg_cb(uint8_t sender_id __attribute__((unused)), 
+	uint8_t ac_id, int8_t source_strength, int8_t rssi)
+{ 
+	int i = -1; // Initialize the index of all tracked drones (-1 for null assumption of no drone found).
+
+	// Check if a new aircraft ID is present, if it's a new ID we start a new EKF for it.
+	if (( !array_find_int(NUAVS-1, IDarray, ac_id, &i))  // If yes, a new drone is found.
+		   && (nf < NUAVS-1))  // If yes, the amount of drones does not exceed the maximum.
+	{
+		IDarray[nf] = ac_id; 				// Store ID in an array (logging purposes)
+		srcstrength[nf] = source_strength;  // Store source strength in an array (logging purposes)
+		ekf_filter_new(&ekf[nf]); 			// Initialize an EKF filter for the newfound drone
+
+		// Set up the Q and R matrices and all the rest
+		// Weights are based on:
+		// Coppola et al, "On-board Communication-based Relative Localization for Collision Avoidance in Micro Air Vehicle teams", 2017
+		fmat_scal_mult(EKF_N,EKF_N, ekf[nf].Q, pow(0.5,2.0), ekf[nf].Q);
```

Comment:
when using the math functions from libc, you should use the float version `powf`. Some for other functions like `sinf`, cosf`, `atanf`, `sqrtf` and so on.

Reply:
Also, if this is a constant computation can't you just precompute it? #efficiency 

If you want to keep it clear where this value comes from you can do something like this:
static const float descriptive_variable_name = powf(0.5f, 2.f);